src/Xrefcheck/Progress.hs

```diff
@@ -22,59 +28,127 @@ import Universum
+-- | Increase the number of unfixable errors.
+incProgressUnfixableErrors :: (Num a, Show a) => Progress a -> Progress a
+incProgressUnfixableErrors Progress{..} = Progress{ pErrorsUnfixable = pErrorsUnfixable + 1
+                                                  , ..
+                                                  }
+
+-- | Increase the number of fixable errors.
+incProgressFixableErrors :: (Num a, Show a) => Progress a -> Progress a
+incProgressFixableErrors Progress{..} = Progress{ pErrorsFixable = pErrorsFixable + 1
+                                                , ..
+                                                }
+
+-- | Decrease the number of fixable errors. This function indicates the situation where one of
+-- such errors had been successfully eliminated.
+decProgressFixableErrors :: (Num a, Show a) => Progress a -> Progress a
+decProgressFixableErrors Progress{..} = Progress{ pErrorsFixable = pErrorsFixable - 1
+                                                , ..
+                                                }
+
+fixableToUnfixable :: (Num a, Show a) => Progress a -> Progress a
+fixableToUnfixable Progress{..} = Progress{ pErrorsFixable = pErrorsFixable - 1
+                                          , pErrorsUnfixable = pErrorsUnfixable + 1
+                                          , ..
+                                          }
+
+setTaskTimestamp :: Time Second -> Time Second -> Progress a -> Progress a
+setTaskTimestamp ttc startTime Progress{..} = Progress{ pTaskTimestamp = Just (ttc, startTime)
+                                                      , ..
+                                                      }
+
+removeTaskTimestamp :: Progress a -> Progress a
+removeTaskTimestamp Progress{..} = Progress{ pTaskTimestamp = Nothing
+                                           , ..
+                                           }
+
+checkTaskTimestamp :: Time Second -> Progress a -> Progress a
+checkTaskTimestamp posixTime p@Progress{..} =
+  let mWithinGivenTime = do
+        (ttc, startTime) <- pTaskTimestamp
+        return $ ttc >= posixTime -:- startTime
+  in maybe p (\withinGivenTime ->
+    if withinGivenTime
+    then p
+    else removeTaskTimestamp p) mWithinGivenTime
 
 -- | Visualise progress bar.
-showProgress :: Text -> Int -> Color -> Progress Int -> Text
-showProgress name width col Progress{..} = mconcat
+showProgress :: Text -> Int -> Color -> Time Second -> Progress Int -> Text
+showProgress name width col posixTime Progress{..} = mconcat
   [ color col (name <> ": [")
   , toText bar
+  , timer
   , color col "]"
   , status
   ]
   where
     done = floor $ (pCurrent % pTotal) * fromIntegral width
-    errs = ceiling $ (pErrors % pTotal) * fromIntegral width
-    done' = max 0 $ done - errs
-    remained' = width - errs - done'
+    errsU = ceiling $ (pErrorsUnfixable % pTotal) * fromIntegral width
+    errsF = min (width - errsU) . ceiling $ (pErrorsFixable % pTotal) * fromIntegral width
```

Comment:
Regarding your [comment here](https://github.com/serokell/xrefcheck/pull/88#issuecomment-965740813): yup, your explanation and your solution make sense, well done! :+1: 

I would add some comments in the code though, to add some context (otherwise our future selves will have to spend a full hour scratching their heads trying to figure out what this means :sweat_smile: )

In particular, please document why (1) we're using `ceiling` in the first place and (2) that, due to our use of `ceiling`, the progress bar may overflow and thus we need to use `min` to prevent this.

While we're here, might as well explain why we're using `max 0` in the line below too.